File: implementation/scipamato/scipamato-core-sync/src/test/java/ch/difty/scipamato/core/sync/launcher/RefDataSyncJobLauncherTest.java
package ch.difty.scipamato.core.sync.launcher;

import static java.util.stream.Collectors.toList;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.*;

import java.util.*;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.batch.core.*;
import org.springframework.batch.core.launch.JobLauncher;
import org.springframework.batch.core.repository.JobExecutionAlreadyRunningException;
import org.springframework.batch.core.repository.JobInstanceAlreadyCompleteException;
import org.springframework.batch.core.repository.JobRestartException;

@SuppressWarnings("SameParameterValue")
@ExtendWith(MockitoExtension.class)
public class RefDataSyncJobLauncherTest {

    private static final int    JOB_STEP_ID_START    = 75;
    private static final int    BATCH_SIZE           = 100;
    private static final String UNSYNCHED_PAPERS_MSG = "unsynched papers";

    private RefDataSyncJobLauncher launcher;

    @Mock
    private JobLauncher jobLauncher;
    @Mock
    private Warner      warner;

    @Mock
    private Job syncLanguageJob, syncNewStudyPageLinkJob, syncCodeClassJob, syncCodeJob, syncPaperJob;
    @Mock
    private Job syncNewsletterJob, syncNewsletterTopicJob, syncNewStudyJob, syncNewStudyTopicJob, syncKeywordJob;

    private Map<String, Job> jobMap;

    @BeforeEach
    public void setUp() {
        launcher = new RefDataSyncJobLauncher(jobLauncher, syncLanguageJob, syncNewStudyPageLinkJob, syncCodeClassJob,
            syncCodeJob, syncPaperJob, syncNewsletterJob, syncNewsletterTopicJob, syncNewStudyJob, syncNewStudyTopicJob,
            syncKeywordJob, warner);

        jobMap = jobsPerTopic();
    }

    @AfterEach
    public void tearDown() {
        verifyNoMoreInteractions(jobLauncher, syncLanguageJob, syncNewStudyPageLinkJob, syncCodeClassJob, syncCodeJob,
            syncPaperJob, syncNewsletterJob, syncNewsletterTopicJob, syncNewStudyJob, syncNewStudyTopicJob,
            syncKeywordJob, warner);
    }

    private Map<String, Job> jobsPerTopic() {
        final Map<String, Job> jobMap = new LinkedHashMap<>();
        jobMap.put("languages", syncLanguageJob);
        jobMap.put("newStudyPage_links", syncNewStudyPageLinkJob);
        jobMap.put("code_classes", syncCodeClassJob);
        jobMap.put("codes", syncCodeJob);
        jobMap.put("papers", syncPaperJob);
        jobMap.put("newsletters", syncNewsletterJob);
        jobMap.put("newsletterTopics", syncNewsletterTopicJob);
        jobMap.put("newStudyTopics", syncNewStudyTopicJob);
        jobMap.put("newStudies", syncNewStudyJob);
        jobMap.put("keywords", syncKeywordJob);
        return jobMap;
    }

    @Test
    public void jobParameters_haveSingleIdentifyingParameterRunDate_withCurrentDate() {
        JobParameters params = launcher.getJobParameters();
        assertThat(params.getParameters()).hasSize(1);
        assertThat(params.getDate("runDate")).isCloseTo(new Date(), 1000);
        assertThat(params
            .getParameters()
            .values()
            .iterator()
            .next()
            .isIdentifying()).isTrue();
    }

    @Test
    public void launching_withUnsynchronizedPapersAndAllStepsSuccessful_addsWarningBeforeStepResultsAndSucceeds()
        throws JobParametersInvalidException, JobExecutionAlreadyRunningException, JobRestartException,
        JobInstanceAlreadyCompleteException {

        when(warner.findUnsynchronizedPapers()).thenReturn(Optional.of(UNSYNCHED_PAPERS_MSG));
        final List<String> expectedMessages = messagesWithAllStepsSuccessful(jobMap, true);

        final SyncJobResult result = launcher.launch();

        assertThat(result.isSuccessful()).isTrue();
        assertThat(result.isFailed()).isFalse();
        assertAllJobsSuccessfulButWithUnsynchedPapers(expectedMessages, result);

        verifyMocks(jobMap);
    }

    private List<String> messagesWithAllStepsSuccessful(final Map<String, Job> jobMap, boolean withUnsynchedPapers)
        throws JobExecutionAlreadyRunningException, JobRestartException, JobInstanceAlreadyCompleteException,
        JobParametersInvalidException {

        final List<String> expectedMessages = new ArrayList<>();

        if (withUnsynchedPapers)
            expectedMessages.add(UNSYNCHED_PAPERS_MSG);

        int jobId = JOB_STEP_ID_START;
        for (final Map.Entry<String, Job> entry : jobMap.entrySet()) {
            // simple fixture using jobId to get variance in records written
            final int writtenRecords = jobId + BATCH_SIZE;
            final String msg = String.format(
                "Job %d has returned with exitCode COMPLETED (status COMPLETED): %d %s were synchronized.", jobId,
                writtenRecords, entry.getKey());
            expectedMessages.add(msg);
            jobId++;
        }

        jobId = JOB_STEP_ID_START;
        for (final Map.Entry<String, Job> entry : jobMap.entrySet()) {
            jobLauncherFixture(jobId++, BatchStatus.COMPLETED, ExitStatus.COMPLETED, entry.getValue());
        }
        return expectedMessages;
    }

    private void jobLauncherFixture(final long executionID, final BatchStatus status, final ExitStatus exitStatus,
        final Job job)
        throws JobExecutionAlreadyRunningException, JobRestartException, JobInstanceAlreadyCompleteException,
        JobParametersInvalidException {
        final JobExecution jobExecution = jobExecutionFixture(executionID, status, exitStatus);
        when(jobLauncher.run(eq(job), isA(JobParameters.class))).thenReturn(jobExecution);
    }

    private JobExecution jobExecutionFixture(final long id, final BatchStatus status, final ExitStatus exitStatus) {
        final JobExecution jobExecution = mock(JobExecution.class);
        when(jobExecution.getId()).thenReturn(id);
        when(jobExecution.getStatus()).thenReturn(status);
        when(jobExecution.getExitStatus()).thenReturn(exitStatus);

        StepExecution stepExecution1 = mock(StepExecution.class);
        StepExecution stepExecution2 = mock(StepExecution.class);
        when(stepExecution1.getWriteCount()).thenReturn(BATCH_SIZE);
        // simple fixture to get some variance in the returned records
        when(stepExecution2.getWriteCount()).thenReturn((int) id);

        when(jobExecution.getStepExecutions()).thenReturn(List.of(stepExecution1, stepExecution2));

        return jobExecution;
    }

    private void assertAllJobsSuccessfulButWithUnsynchedPapers(final List<String> expectedMessages,
        final SyncJobResult result) {
        assertThat(result.getMessages()).hasSize(11);

        // warning due to unsynchronized papers
        final SyncJobResult.LogMessage logMessage = result
            .getMessages()
            .get(0);
        assertThat(logMessage.getMessage()).isEqualTo(UNSYNCHED_PAPERS_MSG);
        assertThat(logMessage.getMessageLevel()).isEqualTo(SyncJobResult.MessageLevel.WARNING);

        // job step results
        assertThat(result
            .getMessages()
            .subList(1, result
                .getMessages()
                .size()))
            .extracting("messageLevel")
            .containsOnly(SyncJobResult.MessageLevel.INFO);
        assertThat(result.getMessages())
            .extracting("message")
            .containsExactlyElementsOf(expectedMessages);
    }

    private void assertAllJobsSuccessfulWithNoUnsynchedPapers(final List<String> expectedMessages,
        final SyncJobResult result) {
        assertThat(result.getMessages()).hasSize(10);
        assertThat(result.getMessages())
            .extracting("messageLevel")
            .containsOnly(SyncJobResult.MessageLevel.INFO);
        assertThat(result.getMessages())
            .extracting("message")
            .containsExactlyElementsOf(expectedMessages);
    }

    private void verifyMocks(final Map<String, Job> jobMap)
        throws JobExecutionAlreadyRunningException, JobRestartException, JobInstanceAlreadyCompleteException,
        JobParametersInvalidException {
        verify(warner).findUnsynchronizedPapers();
        for (final Map.Entry<String, Job> entry : jobMap.entrySet()) {
            verify(jobLauncher).run(eq(entry.getValue()), isA(JobParameters.class));
        }
    }

    @Test
    public void launching_withoutUnsynchronizedPapers_onlyAddsInfoMessages()
        throws JobParametersInvalidException, JobExecutionAlreadyRunningException, JobRestartException,
        JobInstanceAlreadyCompleteException {
        when(warner.findUnsynchronizedPapers()).thenReturn(Optional.empty());
        final List<String> expectedMessages = messagesWithAllStepsSuccessful(jobMap, false);

        final SyncJobResult result = launcher.launch();
        assertThat(result.isSuccessful()).isTrue();

        assertAllJobsSuccessfulWithNoUnsynchedPapers(expectedMessages, result);

        verifyMocks(jobMap);
    }

    @Test
    public void launching_withFailingStep_failsJob()
        throws JobParametersInvalidException, JobExecutionAlreadyRunningException, JobRestartException,
        JobInstanceAlreadyCompleteException {
        when(warner.findUnsynchronizedPapers()).thenReturn(Optional.empty());
        final List<String> expectedMessages = messagesWithFailingStepInPosition3(jobMap);

        final SyncJobResult result = launcher.launch();
        assertThat(result.isSuccessful()).isFalse();
        assertThat(result.isFailed()).isTrue();

        assertAllJobsSuccessfulExceptThird(expectedMessages, result);

        verifyMocks(jobMap);
    }

    private List<String> messagesWithFailingStepInPosition3(final Map<String, Job> jobMap)
        throws JobExecutionAlreadyRunningException, JobRestartException, JobInstanceAlreadyCompleteException,
        JobParametersInvalidException {
        final int failAfter = 2;

        final List<String> expectedMessages = new ArrayList<>();
        int jobId = JOB_STEP_ID_START;
        for (final Map.Entry<String, Job> entry : jobMap.entrySet()) {
            // simple fixture using jobId to get variance in records written
            final int writtenRecords = jobId + BATCH_SIZE;
            if (jobId != JOB_STEP_ID_START + failAfter) {
                final String msg = String.format(
                    "Job %d has returned with exitCode COMPLETED (status COMPLETED): %d %s were synchronized.", jobId,
                    writtenRecords, entry.getKey());
                expectedMessages.add(msg);
            } else {
                final String msg = String.format(
                    "Job %d has returned with exitCode FAILED (status FAILED): %d %s were synchronized.", jobId,
                    writtenRecords, entry.getKey());
                expectedMessages.add(msg);
            }
            jobId++;
        }

        jobId = JOB_STEP_ID_START;
        for (final Map.Entry<String, Job> entry : jobMap.entrySet()) {
            if (jobId != JOB_STEP_ID_START + failAfter) {
                jobLauncherFixture(jobId++, BatchStatus.COMPLETED, ExitStatus.COMPLETED, entry.getValue());
            } else {
                jobLauncherFixture(jobId++, BatchStatus.FAILED, ExitStatus.FAILED, entry.getValue());
            }
        }
        return expectedMessages;
    }

    private void assertAllJobsSuccessfulExceptThird(final List<String> expectedMessages, final SyncJobResult result) {
        assertThat(result.getMessages()).hasSize(10);
        assertThat(result.getMessages())
            .extracting("messageLevel")
            .containsExactly(SyncJobResult.MessageLevel.INFO, SyncJobResult.MessageLevel.INFO,
                SyncJobResult.MessageLevel.ERROR, SyncJobResult.MessageLevel.INFO, SyncJobResult.MessageLevel.INFO,
                SyncJobResult.MessageLevel.INFO, SyncJobResult.MessageLevel.INFO, SyncJobResult.MessageLevel.INFO,
                SyncJobResult.MessageLevel.INFO, SyncJobResult.MessageLevel.INFO);
        assertThat(result.getMessages())
            .extracting("message")
            .containsExactlyElementsOf(expectedMessages);
    }

    @Test
    public void launching_withUnexpectedException_stopsRunningSubsequentJobs()
        throws JobParametersInvalidException, JobExecutionAlreadyRunningException, JobRestartException,
        JobInstanceAlreadyCompleteException {
        when(warner.findUnsynchronizedPapers()).thenReturn(Optional.empty());
        final List<String> expectedMessages = messagesWithExceptionAfter2nd(jobMap);
        expectedMessages.add(
            "Unexpected exception of type class java.lang.RuntimeException: unexpected exception somewhere");

        final SyncJobResult result = launcher.launch();
        assertThat(result.isSuccessful()).isFalse();
        assertThat(result.isFailed()).isTrue();

        assertThat(result.getMessages())
            .extracting("messageLevel")
            .containsExactly(SyncJobResult.MessageLevel.INFO, SyncJobResult.MessageLevel.INFO,
                SyncJobResult.MessageLevel.ERROR);
        assertThat(result.getMessages())
            .extracting("message")
            .containsExactlyElementsOf(expectedMessages);

        verify(warner).findUnsynchronizedPapers();
        for (final Map.Entry<String, Job> entry : jobMap
            .entrySet()
            .stream()
            .limit(3)
            .collect(toList())) {
            verify(jobLauncher).run(eq(entry.getValue()), isA(JobParameters.class));
        }
    }

    private List<String> messagesWithExceptionAfter2nd(final Map<String, Job> jobMap)
        throws JobExecutionAlreadyRunningException, JobRestartException, JobInstanceAlreadyCompleteException,
        JobParametersInvalidException {
        final int failAfter = 2;

        final List<String> expectedMessages = new ArrayList<>();
        int jobId = JOB_STEP_ID_START;
        for (final Map.Entry<String, Job> entry : jobMap.entrySet()) {
            // simple fixture using jobId to get variance in records written
            final int writtenRecords = jobId + BATCH_SIZE;
            final String msg = String.format(
                "Job %d has returned with exitCode COMPLETED (status COMPLETED): %d %s were synchronized.", jobId,
                writtenRecords, entry.getKey());
            expectedMessages.add(msg);
            jobId++;
            if (jobId == JOB_STEP_ID_START + failAfter)
                break;
        }

        jobId = JOB_STEP_ID_START;
        for (final Map.Entry<String, Job> entry : jobMap.entrySet()) {
            jobLauncherFixture(jobId++, BatchStatus.COMPLETED, ExitStatus.COMPLETED, entry.getValue());
            if (jobId == JOB_STEP_ID_START + failAfter)
                break;
        }

        when(jobLauncher.run(eq(syncCodeClassJob), isA(JobParameters.class))).thenThrow(
            new RuntimeException("unexpected exception somewhere"));
        return expectedMessages;
    }
}

File: core/core-sync/src/test/kotlin/ch/difty/scipamato/core/sync/launcher/RefDataSyncJobLauncherTest.kt
@file:Suppress("SpellCheckingInspection")

package ch.difty.scipamato.core.sync.launcher

import io.mockk.confirmVerified
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.amshove.kluent.shouldBeEmpty
import org.amshove.kluent.shouldBeEqualTo
import org.amshove.kluent.shouldBeTrue
import org.amshove.kluent.shouldContainSame
import org.amshove.kluent.shouldHaveSize
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.batch.core.BatchStatus
import org.springframework.batch.core.ExitStatus
import org.springframework.batch.core.Job
import org.springframework.batch.core.JobExecution
import org.springframework.batch.core.StepExecution
import org.springframework.batch.core.launch.JobLauncher

internal class RefDataSyncJobLauncherTest {

    private val jobLauncher = mockk<JobLauncher>()
    private val warner = mockk<Warner>()

    private val syncLanguageJob = mockk<Job>()
    private val syncNewStudyPageLinkJob = mockk<Job>()
    private val syncCodeClassJob = mockk<Job>()
    private val syncCodeJob = mockk<Job>()
    private val syncPaperJob = mockk<Job>()
    private val syncNewsletterJob = mockk<Job>()
    private val syncNewsletterTopicJob = mockk<Job>()
    private val syncNewStudyJob = mockk<Job>()
    private val syncNewStudyTopicJob = mockk<Job>()
    private val syncKeywordJob = mockk<Job>()

    private var launcher = RefDataSyncJobLauncher(
        jobLauncher, syncLanguageJob, syncNewStudyPageLinkJob,
        syncCodeClassJob, syncCodeJob, syncPaperJob, syncNewsletterJob, syncNewsletterTopicJob, syncNewStudyJob,
        syncNewStudyTopicJob, syncKeywordJob, warner
    )

    private val success: MutableList<String> = mutableListOf()
    private val failure: MutableList<String> = mutableListOf()
    private val warning: MutableList<String> = mutableListOf()
    private var done: Boolean = false

    private val setSuccess: (String) -> Unit = { success.add(it) }
    private val setFailure: (String) -> Unit = { failure.add(it) }
    private val setWarning: (String) -> Unit = { warning.add(it) }
    private val setDone: () -> Unit = { done = true }

    private var jobMap = jobsPerTopic()

    @BeforeEach
    fun setUp() {
        success.clear()
        failure.clear()
        warning.clear()
        done = false
    }

    @AfterEach
    fun tearDown() {
        confirmVerified(jobLauncher, warner)
    }

    private fun jobsPerTopic(): Map<String, Job> {
        return mapOf(
            "languages" to syncLanguageJob,
            "newStudyPage_links" to syncNewStudyPageLinkJob,
            "code_classes" to syncCodeClassJob,
            "codes" to syncCodeJob,
            "papers" to syncPaperJob,
            "newsletters" to syncNewsletterJob,
            "newsletterTopics" to syncNewsletterTopicJob,
            "newStudyTopics" to syncNewStudyTopicJob,
            "newStudies" to syncNewStudyJob,
            "keywords" to syncKeywordJob
        )
    }

    @Test
    fun jobParameters_haveSingleIdentifyingParameterRunDate_withCurrentDate() {
        val params = launcher.jobParameters
        params.parameters.shouldHaveSize(1)
        params.parameters.values.first().isIdentifying.shouldBeTrue()
    }

    @Test
    fun launching_withUnsynchronizedPapersAndAllStepsSuccessful_addsWarningBeforeStepResultsAndSucceeds() {
        every { warner.findUnsynchronizedPapers() } returns UNSYNCHED_PAPERS_MSG
        every { warner.findNewsletterswithUnsynchronizedPapers() } returns null
        val successMsg = messagesWithAllStepsSuccessful(jobMap)

        launcher.launch(setSuccess, setFailure, setWarning, setDone)

        assertAllJobsSuccessfulButWithUnsynchedPapers(successMsg, UNSYNCHED_PAPERS_MSG)

        verifyMocks(jobMap)
    }

    @Test
    fun launching_withUnsynchronizedPapersAndNewslettersAndAllStepsSuccessful_addsWarningBeforeStepResultsAndSucceeds() {
        every { warner.findUnsynchronizedPapers() } returns null
        every { warner.findNewsletterswithUnsynchronizedPapers() } returns UNSYNCHED_NEWSLETTER_MSG
        val successMsg = messagesWithAllStepsSuccessful(jobMap)

        launcher.launch(setSuccess, setFailure, setWarning, setDone)

        assertAllJobsSuccessfulButWithUnsynchedPapers(successMsg, UNSYNCHED_NEWSLETTER_MSG)

        verifyMocks(jobMap)
    }

    private fun messagesWithAllStepsSuccessful(jobMap: Map<String, Job>): List<String> {
        val expectedMessages = ArrayList<String>()

        var jobId = JOB_STEP_ID_START
        jobMap.keys.forEach { key ->
            // simple fixture using jobId to get variance in records written
            val writtenRecords = jobId + BATCH_SIZE
            val msg = "Job $jobId has returned with exitCode COMPLETED (status COMPLETED): " +
                "$writtenRecords $key were synchronized."
            expectedMessages.add(msg)
            jobId++
        }

        jobId = JOB_STEP_ID_START
        jobMap.values.forEach { value ->
            jobLauncherFixture(jobId++.toLong(), BatchStatus.COMPLETED, ExitStatus.COMPLETED, value)
        }

        return expectedMessages
    }

    private fun jobLauncherFixture(executionID: Long, status: BatchStatus, exitStatus: ExitStatus, job: Job) {
        val jobExecution = jobExecutionFixture(executionID, status, exitStatus)
        every { jobLauncher.run(eq(job), any()) } returns jobExecution
    }

    private fun jobExecutionFixture(id: Long, status: BatchStatus, exitStatus: ExitStatus): JobExecution {
        val jobExecution = mockk<JobExecution>()
        every { jobExecution.id } returns id
        every { jobExecution.status } returns status
        every { jobExecution.exitStatus } returns exitStatus

        val stepExecution1 = mockk<StepExecution>()
        val stepExecution2 = mockk<StepExecution>()
        every { stepExecution1.writeCount } returns BATCH_SIZE
        // simple fixture to get some variance in the returned records
        every { stepExecution2.writeCount } returns id

        every { jobExecution.stepExecutions } returns listOf(stepExecution1, stepExecution2)

        return jobExecution
    }

    private fun assertAllJobsSuccessfulButWithUnsynchedPapers(successMsg: List<String>, warnMsg: String) {
        warning.shouldHaveSize(1)
        warning.first() shouldBeEqualTo warnMsg
        success shouldContainSame successMsg
        failure.shouldBeEmpty()
        done.shouldBeTrue()
    }

    private fun assertAllJobsSuccessfulWithNoUnsynchedPapers(expectedMessages: List<String>) {
        success shouldContainSame expectedMessages
        warning.shouldBeEmpty()
        failure.shouldBeEmpty()
        done.shouldBeTrue()
    }

    private fun verifyMocks(jobMap: Map<String, Job>) {
        verify { warner.findUnsynchronizedPapers() }
        verify { warner.findNewsletterswithUnsynchronizedPapers() }
        jobMap.values.forEach { job -> verify { jobLauncher.run(job, any()) } }
    }

    @Test
    fun launching_withoutUnsynchronizedPapers_onlyAddsInfoMessages() {
        every { warner.findUnsynchronizedPapers() } returns null
        every { warner.findNewsletterswithUnsynchronizedPapers() } returns null
        val expectedMessages = messagesWithAllStepsSuccessful(jobMap)

        launcher.launch(setSuccess, setFailure, setWarning, setDone)

        assertAllJobsSuccessfulWithNoUnsynchedPapers(expectedMessages)

        verifyMocks(jobMap)
    }

    @Test
    fun launching_withFailingStep_failsJob() {
        every { warner.findUnsynchronizedPapers() } returns null
        every { warner.findNewsletterswithUnsynchronizedPapers() } returns null
        val (successMsg, failureMsg) = messagesWithFailingStepInPosition3(jobMap)

        launcher.launch(setSuccess, setFailure, setWarning, setDone)

        success shouldContainSame successMsg
        failure shouldContainSame failureMsg
        warning.shouldBeEmpty()

        verifyMocks(jobMap)
    }

    private fun messagesWithFailingStepInPosition3(jobMap: Map<String, Job>): Pair<List<String>, List<String>> {
        val failAfter = 2

        val successMessages = ArrayList<String>()
        val failureMessages = ArrayList<String>()
        var jobId = JOB_STEP_ID_START
        jobMap.keys.forEach { key ->
            // simple fixture using jobId to get variance in records written
            val writtenRecords = jobId + BATCH_SIZE
            if (jobId != JOB_STEP_ID_START + failAfter) {
                val msg = "Job $jobId has returned with exitCode COMPLETED (status COMPLETED): " +
                    "$writtenRecords $key were synchronized."
                successMessages.add(msg)
            } else {
                val msg = "Job $jobId has returned with exitCode FAILED (status FAILED): " +
                    "$writtenRecords $key were synchronized."
                failureMessages.add(msg)
            }
            jobId++
        }

        jobId = JOB_STEP_ID_START
        jobMap.values.forEach { value ->
            if (jobId != JOB_STEP_ID_START + failAfter) {
                jobLauncherFixture(jobId++.toLong(), BatchStatus.COMPLETED, ExitStatus.COMPLETED, value)
            } else {
                jobLauncherFixture(jobId++.toLong(), BatchStatus.FAILED, ExitStatus.FAILED, value)
            }
        }
        return successMessages to failureMessages
    }

    @Test
    fun launching_withUnexpectedException_stopsRunningSubsequentJobs() {
        every { warner.findUnsynchronizedPapers() } returns null
        every { warner.findNewsletterswithUnsynchronizedPapers() } returns null

        val (successMsg, failureMsg) = messagesWithExceptionAfter2nd(jobMap)

        launcher.launch(setSuccess, setFailure, setWarning, setDone)

        success shouldContainSame successMsg
        failure shouldContainSame failureMsg

        verify { warner.findUnsynchronizedPapers() }
        verify { warner.findNewsletterswithUnsynchronizedPapers() }

        jobMap.values.take(3).forEach { job -> verify { jobLauncher.run(job, any()) } }
    }

    @Suppress("ReturnCount")
    private fun messagesWithExceptionAfter2nd(jobMap: Map<String, Job>): Pair<List<String>, List<String>> {
        val failAfter = 2

        val successMsg = ArrayList<String>()
        val failureMsg = ArrayList<String>()
        var jobId = JOB_STEP_ID_START
        run keyLoop@{
            jobMap.keys.forEach { key ->
                // simple fixture using jobId to get variance in records written
                val writtenRecords = jobId + BATCH_SIZE
                val msg = "Job $jobId has returned with exitCode COMPLETED (status COMPLETED): " +
                    "$writtenRecords $key were synchronized."
                successMsg.add(msg)
                if (++jobId == JOB_STEP_ID_START + failAfter) return@keyLoop
            }
        }

        jobId = JOB_STEP_ID_START
        run valueLoop@{
            jobMap.values.forEach { value ->
                jobLauncherFixture(jobId++.toLong(), BatchStatus.COMPLETED, ExitStatus.COMPLETED, value)
                if (jobId == JOB_STEP_ID_START + failAfter) return@valueLoop
            }
        }

        every { jobLauncher.run(syncCodeClassJob, any()) } throws RuntimeException("unexpected exception somewhere")

        failureMsg.add("Unexpected exception of type class java.lang.RuntimeException: unexpected exception somewhere")
        return successMsg to failureMsg
    }

    companion object {
        private const val JOB_STEP_ID_START = 75
        private const val BATCH_SIZE = 100L
        private const val UNSYNCHED_PAPERS_MSG = "unsynched papers"
        private const val UNSYNCHED_NEWSLETTER_MSG = "unsynched newsletters"
    }
}
